extension/background.js

```diff
@@ -0,0 +1,196 @@
+      await chrome.tabs.update(tabId, { url: chrome.runtime.getURL('prompt.html') });
+      await new Promise((resolve) => {
+        const listener = (message, foo) => {
+          if (foo.tab.id === tabId && message.action === 'approve') {
+            chrome.runtime.onMessage.removeListener(listener);
+            resolve(undefined);
+          }
+        };
+        chrome.runtime.onMessage.addListener(listener);
+      });
+    }
+    const debuggee = { tabId }
+    await chrome.debugger.attach(debuggee, '1.3')
+    if (chrome.runtime.lastError) {
+      debugLog('Failed to attach debugger:', chrome.runtime.lastError.message);
+      return;
+    }
+    debugLog('Debugger attached to tab:', debuggee.tabId);
+    const { targetId, browserContextId } = (/** @type{any} */ (await chrome.debugger.sendCommand(debuggee, 'Target.getTargetInfo', {}))).targetInfo;
+    const socket = new WebSocket(proxyURL);
+    const adapter = new CDPAdapter(tabId);
+    adapter.dispatch = (data) => {
+      debugLog('Sending message to browser:', data);
+      if (socket.readyState === WebSocket.OPEN) {
+        socket.send(JSON.stringify(data));
+      } else {
+        debugLog('WebSocket is not open. Cannot send data.');
+      }
+    }
+    adapter.onClose = async () => {
+      debugLog('Debugger detached from tab:', tabId);
+      this.adapters.delete(tabId);
+      if (socket.readyState === WebSocket.OPEN)
+        socket.close();
+    }
+    socket.addEventListener('open', () => {
+      chrome.tabs.update(debuggee.tabId, { url: chrome.runtime.getURL('success.html') });
+    });
+    socket.addEventListener('message', (e) => adapter.onBrowserMessage(targetId, browserContextId, e));
+    socket.addEventListener('error', (event) => {
+      debugLog('WebSocket error:', event);
+      adapter.detach();
+    });
+    socket.addEventListener('close', async () => {
+      adapter.detach();
+    });
+    this.adapters.set(tabId, adapter);
+  }
+}
+
+class CDPAdapter {
+  /**
+   * @param {number} tabId
+   */
+  constructor(tabId) {
+    chrome.debugger.onEvent.addListener((this._onDebuggerEvent.bind(this)));
+    chrome.debugger.onDetach.addListener(this._onDebuggerDetach.bind(this));
+    this.onClose = () => { }
+    this.dispatch = (data) => { }
+    this._debuggee = { tabId };
+  }
+
+  /**
+   * @param {string} targetId 
+   * @param {string} browserContextId 
+   * @param {object} event 
+   * @returns 
+   */
+  async onBrowserMessage(targetId, browserContextId, event) {
+    try {
+      const message = JSON.parse(await event.data.text());
+      if (message.method === 'Browser.getVersion') {
+        // Handle the Browser.getVersion command
+        let versionInfo = {
+          protocolVersion: '1.3',
+          userAgent: navigator.userAgent,
+          product: 'Chrome'
+        };
+        this.dispatch({ id: message.id, result: versionInfo });
+        return;
+      }
+      if (message.method === 'Target.setAutoAttach' && !message.sessionId) {
```

Comment:
If `message.sessionId` exists, I assume it will just forward the request to `chrome.debugger.sendCommand`, right? I had some issues in the past with `Target.setAutoAttach` requests on pages with service workers:

https://github.com/ruifigueira/playwright-crx/blob/eb65d74c6bb0e9f04cc4fdff1570ac9b2c53d0fe/src/server/transport/crxTransport.ts#L89-L101